### PR TITLE
Allow custom camera app to be specified in a question

### LINKED ISSUE #5621: Allow custom camera app to be specified in a question
- [ ] Given I tap the `Take` button for an image question with a custom app specified through the `intent` body attribute
  And I have the custom app on my device
  Then I see the custom app

- [ ] Given I tap the `Take` button for an image question with a custom app specified
  And I don't have the custom app on my device or the app I specified doesn't respond to `ACTION_IMAGE_CAPTURE`
  Then I see an error toast (same as what [external widgets](https://docs.getodk.org/launch-apps-from-collect/) do)

- [ ] Given I took a picture with my custom camera app after launching it from the `Take` button on an image question
  Then I return to Collect and see the picture I took

- [ ] Given I am at an image question with both a custom app and the `new` appearance specified
  Then I do not see the `Choose` button

Testing considerations:
- Android 12+
- `annotate` should also combine with both this and `new`

Non-goals:
- We're not currently doing a bottom sheet to pick the app: https://github.com/getodk/collect/issues/5621#issuecomment-1645520802; we might later consider a special value for the `intent` attribute like `picker` to let a user pick an app
- We're not doing video because no one has asked for it

## Background
Prior to Android 11 users were allowed to select a camera app they wanted to use to capture image/record video. Then it was changed and now when using `ACTION_IMAGE_CAPTURE` action the default (built-in) camera app is started. This is a problem for users that want to use specific camera apps.
It turns out that if we know the package name of an app we can force starting it by adding it to the intent, eg.:

`            intent.setPackage("com.jeyluta.timestampcamerafree");`

I gave it a try and it works well with the[ Timestamp camera app](https://play.google.com/store/apps/details?id=com.jeyluta.timestampcamerafree&hl=en&gl=US) 

If it's possible to do the same with other apps too we can allow adding package names to the [ImageWidget](https://docs.getodk.org/form-question-types/#image-widgets) as an `appearance` (like in https://docs.getodk.org/form-question-types/#external-app-image-widget) or `parameter`.

P.S.
Adding queries in the manifest file might also be required:

```
<queries>
    <intent>
        <action android:name="android.media.action.IMAGE_CAPTURE" />
    </intent>
</queries>
```

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #5621 

#### What has been done to verify that this works as intended?
I've tested the changes manually and added automated tests.

#### Why is this the best possible solution? Were any other approaches considered?
We have discussed other options (see in the issue) but decided that the issue should be fixed in this particular way.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
As described in the issue this should allow specifying custom camera apps in image/annotate widgets. We need to test carefully those two widgets and verify edge cases like:
- what happens if a widget has a custom camera app specified but it's not installed 
- what if the package name is wrong
- what if the package name is correct but points to an app tat is not a camera app

#### Do we need any specific form for testing your changes? If so, please attach one.
I've used:
[image_questions.xlsx](https://github.com/getodk/collect/files/13249701/image_questions.xlsx)
[image_questions.xml.txt](https://github.com/getodk/collect/files/13249712/image_questions.xml.txt)

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
Yes https://github.com/getodk/docs/issues/1652

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

**Comment by lognaturel** [comment]
@seadowg will consider @lindsay-stevens's points at https://github.com/XLSForm/pyxform/issues/651#issuecomment-1730839902 when doing review. @grzesiek2010 and I still feel like it's the right way to go.

**Comment by seadowg** [comment]
@lognaturel @grzesiek2010 I don't actually think it makes sense for me to jump into XLSForm/XForm design discussion here if you're both already convinced we're going in the right direction - I think I'm hitting my context switching limit. If there's still some doubt that'd be helpful for me to weight in on though, let me know and I can dive in. For the moment, I'll remove myself as a reviewer and you can add me back on if this is ready to go.

**Comment by grzesiek2010** [comment]
I think it's ready for review. We decided to stick to what we agreed on before and changes to pyxform will be merged soon.

**Comment by seadowg** [comment]
@grzesiek2010 Would it be possible to include an example XLSForm (and XForm seeing as the pyxform changes won't have released yet) in the PR description. Would make reviewing/testing a lot easier.

**Review by seadowg — CHANGES_REQUESTED** [review]
https://github.com/getodk/collect/pull/5738#issuecomment-1785244525

**Comment by grzesiek2010** [comment]
> @grzesiek2010 Would it be possible to include an example XLSForm (and XForm seeing as the pyxform changes won't have released yet) in the PR description. Would make reviewing/testing a lot easier.

Yes, I forgot to add it.

File: collect_app/src/main/java/org/odk/collect/android/formentry/media/PromptAutoplayer.java [hunk 1]
```diff
@@ -30,7 +31,7 @@ public PromptAutoplayer(AudioHelper audioHelper, ReferenceManager referenceManag
     }
 
     public Boolean autoplayIfNeeded(FormEntryPrompt prompt) {
-        String autoplayOption = prompt.getFormElement().getAdditionalAttribute(null, AUTOPLAY_ATTRIBUTE);
+        String autoplayOption = FormEntryPromptUtils.getBodyAttribute(prompt, AUTOPLAY_ATTRIBUTE);
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/formentry/media/PromptAutoplayer.java]: Maybe we should rename this helper to `getAdditionalAttribute` to closer match the JavaRosa API. I guess an alternative would be to rename/alias in JavaRosa as "additional" doesn't feel very clear here.

File: collect_app/src/main/java/org/odk/collect/android/widgets/AnnotateWidget.java [hunk 2]
```diff
@@ -162,27 +156,7 @@ private int calculateScreenOrientation() {
...
-                    new File(tmpImageFilePath));
-            // if this gets modified, the onActivityResult in
-            // FormEntyActivity will also need to be updated.
-            intent.putExtra(android.provider.MediaStore.EXTRA_OUTPUT, uri);
-            FileUtils.grantFilePermissions(intent, uri, getContext());
-        } catch (IllegalArgumentException e) {
-            Timber.e(e);
-        }
-
+        Intent intent = ImageCaptureIntentCreator.INSTANCE.imageCaptureIntent(getFormEntryPrompt(), getContext(), tmpImageFilePath);
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/widgets/AnnotateWidget.java]: You can use `@JvmStatic` to avoid the `.INSTANCE` here.

File: collect_app/src/test/java/org/odk/collect/android/widgets/AnnotateWidgetTest.java [hunk 3]
```diff
@@ -106,6 +107,23 @@ public void buttonsShouldLaunchCorrectIntents() {
         assertExtraEquals(DrawActivity.OPTION, DrawActivity.OPTION_ANNOTATE, intent);
     }
 
+    @Test
+    public void buttonsShouldLaunchCorrectIntentsWhenCustomPackageIsSet() {
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/widgets/AnnotateWidgetTest.java:105]: Instead of testing this in both the Annotate and Image widget classes, you could write one set for `ImageCaptureIntentCreator.imageCaptureIntent`.

**Review by seadowg — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/formentry/media/PromptAutoplayer.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formentry/media/PromptAutoplayer.java]: Done.

File: collect_app/src/main/java/org/odk/collect/android/widgets/AnnotateWidget.java (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/widgets/AnnotateWidget.java]: Done.

File: collect_app/src/test/java/org/odk/collect/android/widgets/AnnotateWidgetTest.java (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/widgets/AnnotateWidgetTest.java:105]: Yes but then I would need tests in both widgets to make sure that `ImageCaptureIntentCreator` is used. To do that I would need to pass it to those widgets (not use as a static method like now) to make it testable.
And also we have other tests that verify intents in those classes so I think the ones I added are similar and could be there.

**Review by seadowg — APPROVED** [review]

**Comment by dbemke** [comment]
After taking a photo in vertical position via Open Camera and Timestamp camera, the photo is shown in Collect in horizontal position. The issue doesn't always occur. Reproduced on Android 10 and 13 (the form from the PR). 
The screenshot after taking a photo in a vertical position (shown in horizontal).
vertical

**Comment by grzesiek2010** [comment]
> After taking a photo in vertical position via Open Camera and Timestamp camera, the photo is shown in Collect in horizontal position. The issue doesn't always occur. Reproduced on Android 10 and 13 (the form from the PR).
The screenshot after taking a photo in a vertical position (shown in horizontal).

You can file issues like that separately after merging this pull request. It might be on our side but it also might be a bug in that external camera app. This is not directly related to letting users start particular camera apps.

**Comment by dbemke** [comment]
In Image and Annotate widget with "Choose image" when there's a wrong package or the external app isn't installed there's a toast on tapping "Take picture" but it's possible to add a photo via "Choose Image" button. I'm not sure what's the expected - "Choose Image" should be enabled or disabled?

**Comment by grzesiek2010** [comment]
> In Image and Annotate widget with "Choose image" when there's a wrong package or the external app isn't installed there's a toast on tapping "Take picture" but it's possible to add a photo via "Choose Image" button. I'm not sure what's the expected - "Choose Image" should be enabled or disabled?

This is expected we were thinking about hiding that button entirely and only allowing users to take new pictures but if that's users want they can mix it with the `new` appearance (see: https://docs.getodk.org/form-question-types/#image-widget-with-required-new-image)

**Comment by srujner** [comment]
The Image Google Camera have a different behavior:

On Android 10:
- "No activity found to handle: Take picture" toast is visible

On Android 13:
- ODK opens built-in android camera app
Screenshot_20231110-142126

**Comment by grzesiek2010** [comment]
> The Image Google Camera have a different behavior:

Are you sure that on both devices it's the same camera app? And the package name is correct? You should be able to verify the package name using https://play.google.com/store/apps/details?id=com.csdroid.pkg. What about other camera apps on that Android 10?

**Comment by srujner** [comment]
> > The Image Google Camera have a different behavior:
> 
> Are you sure that on both devices it's the same camera app? And the package name is correct? You should be able to verify the package name using https://play.google.com/store/apps/details?id=com.csdroid.pkg. What about other camera apps on that Android 10?

Now i see that Android 13 (Google Pixel 6a) is using app called Pixel Camera (it is built-in app), probably it is not available on Redmi 10

**Comment by grzesiek2010** [comment]
> The Image Google Camera have a different behavior:

This indicates that either the app is not installed or the package name is incorrect. But in this case the toast should be displayed in both cases. If the default camera app is started it might mean there is no package name specified at all. Are you sure that the question has it?

**Comment by grzesiek2010** [comment]
> Now i see that Android 13 (Google Pixel 6a) is using app called Pixel Camera (it is built-in app), probably it is not available on Redmi 10

It doesn't matter what the default camera app is. If you have a question with a package name if it's incorrect the built-in camera app should not be started and there should be a toast in both cases. Maybe you have two different forms?

**Comment by srujner** [comment]
> > Now i see that Android 13 (Google Pixel 6a) is using app called Pixel Camera (it is built-in app), probably it is not available on Redmi 10
> 
> It doesn't matter what the default camera app is. If you have a question with a package name if it's incorrect the built-in camera app should not be started and there should be a toast in both cases. Maybe you have two different forms?

We've got the correct Package name in that case and we're using exactly the same form. 
Android 10 is showing an error because the app is not installed (and the same app as in Google Pixel is not available on Redmi)

**Comment by grzesiek2010** [comment]
> Android 10 is showing an error because the app is not installed

That's ok

> (and the same app as in Google Pixel is not available on Redmi)

it shouldn't matter... you have a question with `com.google.android.GoogleCamera` and it's not installed on both devices. Does the default camera on Android 13 have the same package name? That would explain why it's started but it shouldn't happen I guess.

**Comment by srujner** [comment]
> > Android 10 is showing an error because the app is not installed
> 
> That's ok
> 
> > (and the same app as in Google Pixel is not available on Redmi)
> 
> it shouldn't matter... you have a question with `com.google.android.GoogleCamera` and it's not installed on both devices. Does the default camera on Android 13 have the same package name? That would explain why it's started but it shouldn't happen I guess.

I checked package name using the app that You suggested and it turned out that it was actually the default app on Google Pixel.

So to sum up: On Android 13 device the app with the com.google.android.GoogleCamera package name was already installed that's why it was available and on Android 10 device the app with the same package was not installed that's why we've seen the error toast.

**Comment by dbemke** [comment]
In the Image and Annotate widget with OpenCamera there's an option to record a video. It's possible to record a video in the Image widget but the file won't be saved (there's no crash).
opencamera

**Comment by grzesiek2010** [comment]
Triggering that action I set a flag that indicates it should be for image capturing. If there are apps that do not respect it that's not our fault.

**Comment by dbemke** [comment]
**Tested with Success!**

Verified on device with Android 8.1, 10

**Verified cases:**
- criteria in #5621
- xml forms (Image and Annotate widget) with and without custom camera specified in the form
- packages: OpenCamera, Timestamp camera, Google camera, Camera FV-5 Lite, Cymera
- using annotate and image widget with a custom camera specified when the camera app is installed and not
- wrong package in the form
- correct name of a package the app isn’t a camera app (Youtube)
- camera permissions granted/ not granted in custom cameras
- taking, removing, editing photos via a custom camera
- image widget without a custom camera specified in the form when other camera apps are installed and not installed
- Image and Annotate widget with appearance new
- regression checks in the Image widget

**Comment by srujner** [comment]
**Tested with Success!**

Verified on device with Android 12, 13
